### DISCUSSION TITLE
Zero Power Paths Thresholding

### DISCUSSION THREAD

**Discussion by Priyanshjain24** [NVlabs/sionna#475] [Q&A] [answered]
I am working on the `simple_street_canyon_with_cars` ray tracing scenario (with a single car). After computing paths, I found that the squared sum of path loss coefficients, many paths have 0 or very low power. I want to do position based time evolution of Doppler shift calculations which are very compute expensive. So, I want to do the computations on valid paths only (non zero or some thresholded paths). I have been able to find the valid path indices from paths.cir() but now I have to manually remove the non useful paths from all the attributes. Updating the paths variable somehow would be great as I can even visualise the scene with the valid paths. What I'm doing currently:

```Python
cir = paths.cir()[0][0][0][0][0][0]
path_powers = np.sum(np.abs(cir.numpy())**2, axis=-1)
thresh = 0e-10
valid_indices = path_powers > thresh

paths.tau = tf.boolean_mask(paths.tau, valid_indices, axis=3)
paths.theta_t = tf.boolean_mask(paths.theta_t, valid_indices, axis=3)
paths.theta_r = tf.boolean_mask(paths.theta_r, valid_indices, axis=3)
paths.phi_t = tf.boolean_mask(paths.phi_t, valid_indices, axis=3)
paths.phi_r = tf.boolean_mask(paths.phi_r, valid_indices, axis=3)
paths.a = tf.boolean_mask(paths.a, valid_indices, axis=-2)
paths.vertices = tf.boolean_mask(paths.vertices, valid_indices, axis=-2)
paths.objects = tf.boolean_mask(paths.objects, valid_indices, axis=3)
paths.mask = tf.boolean_mask(paths.mask, valid_indices, axis=3)
```
This helps in removing a lot of paths. All paths:
![Screenshot from 2024-06-11 15-38-43](https://github.com/NVlabs/sionna/assets/101460257/2c2a0c04-7562-485e-bcf2-b97776803c98)

After Thresholding:
![Screenshot from 2024-06-11 15-39-37](https://github.com/NVlabs/sionna/assets/101460257/13b45505-2e13-407e-828f-551a6751ac30)

This is very brute force and wanted to know if there is some better way of doing this in Sionna? Any help would be greatly appreciated.

**Comment by faycalaa** [comment, marked as answer]
Hi @Priyanshjain24  ,

Sionna does not provide a feature to discard paths which path loss is below a specified threshold.

It looks like the paths with zero energy are the scattered ones. Could it be that the scattering coefficients of the materials are set to 0? This would lead to the scattered paths carrying no energy. If you don't need scattering, you could then just disable it by setting the `scattering` parameter of `compute_paths()` to `False`.

**Comment by Priyanshjain24** [comment]
You're right. The scattering coefficients were 0. I've changed them to non zero values to make them non zero energy paths. Thanks.
